### ISSUE TITLE
Feature request: Permanent local storage/Same-node StatefulSet rescheduling

### ISSUE THREAD

**Issue by CallMeFoxie** [kubernetes/kubernetes#42725]
Imagine scenario:
cluster of 200 servers, one app (results from them are aggregated upstream), each works on a different set of data which has to be downloaded **locally** before use. Application controlled by Deployment/StatefulSet. During update each instance gets terminated and restarted. There's no direct way to make sure that the updated instance schedules on the same node to re-use already downloaded data from before. (And ideally fallbacks to another node if the old one died)

Or well, just any database that needs local permanent storage and you want to rolling-update it (deployment/RC/StatefulSet)

Possible solutions that I can think of:
fieldRefs in nodeAffinity (reference metadata.name in node's tags?)
hostPath in PV(C) with nodeAffinity/nodeSelector (similar to what https://github.com/kubernetes/kubernetes/issues/9802 talked about but with affinities?)
???

Basically make StatefulSets also have StatefulStorage one could say?

Welcome to discussion and comments, it would really ease our updates.

**Comment by TamalSaha** [comment]
@CallMeFoxie I think https://github.com/kubernetes/kubernetes/issues/42677 is related to this issue.

**Comment by calebamiles** [comment]
Thanks for pointing out a possibly related issue, @TamalSaha. Could someone from @kubernetes/sig-apps-feature-requests and @kubernetes/sig-storage-feature-requests help qualify this feature request? Thanks!

**Comment by bgrant0607** [comment]
Please see:
https://github.com/kubernetes/kubernetes/issues/7562#issuecomment-281469904
https://github.com/kubernetes/community/pull/306
